aesd-char-driver/main.c

```diff
@@ -228,12 +229,147 @@ ssize_t aesd_write(struct file *filp, const char __user *buf, size_t count,
 
 
 
+/**
+ * @brief Custom llseek implementation to support seek operations
+ *
+ * @param filp     File pointer
+ * @param off      Offset value
+ * @param whence   Positional type (SEEK_SET, SEEK_CUR, SEEK_END)
+ * @return New file position on success, or appropriate error code
+ */
+loff_t aesd_llseek(struct file *filp, loff_t off, int whence)
+{
+    struct aesd_dev *dev = filp->private_data;
+    loff_t new_pos = 0;
+    size_t total_size = 0;
+    struct aesd_buffer_entry *entry;
+    uint8_t index;
+
+    if (mutex_lock_interruptible(&dev->mutex_lock))
+    {
+    	printk("mutex lock unsuccess\n");
+        return -ERESTARTSYS;
+    }
+
+    /*switch (whence)
+    {
+    case SEEK_SET:
+    	printk("seek set, %u\n", new_pos);
+        new_pos = off;
+        break;
+    case SEEK_CUR:
+    	printk("seek cur %u\n", new_pos);
+        new_pos = filp->f_pos + off;
+        break;
+    case SEEK_END:
+        // Calculate the total size of all entries in the circular buffer*/
+        //printk("seek end %u\n", new_pos);
+        AESD_CIRCULAR_BUFFER_FOREACH(entry, &dev->buf, index)
+        {
+            total_size += entry->size;
+        }
+        //new_pos = total_size + off;
+        new_pos = fixed_size_llseek(filp, off, whence, total_size);
```

Comment:
@akka2103 you shouldn't need to hold the lock for the fixed_size_llseek call or fpos access.